integration/src/test/java/org/mule/shutdown/HTTPPersistentConnectionsOnShutdownTestCase.java

```diff
@@ -0,0 +1,242 @@
+
+  private static final int SMALL_TIMEOUT_MILLIS = 300;
+  private static final int POLL_DELAY_MILLIS = 50;
+  private static final String SLOW_PROCESSING_ENDPOINT = "/slow";
+  private static final String FAST_PROCESSING_ENDPOINT = "/fast";
+
+  @Rule
+  public DynamicPort dynamicPort = new DynamicPort("listener.port");
+
+  private ExecutorService stopMuleExecutor;
+
+  @Override
+  protected String getConfigFile() {
+    return "org/mule/shutdown/http-persistent-connections-on-shutdown.xml";
+  }
+
+  @Override
+  protected boolean isGracefulShutdown() {
+    return true;
+  }
+
+  @Before
+  public void setup() {
+    stopMuleExecutor = newSingleThreadExecutor();
+  }
+
+  @After
+  public void tearDown() throws MuleException, InterruptedException {
+    stopMuleExecutor.awaitTermination(1000, MILLISECONDS);
+    stopMuleExecutor.shutdownNow();
+    if (muleContext.isStopped()) {
+      muleContext.start();
+    }
+  }
+
+  @Test
+  public void requestInflightDuringShutdownIsRespondedIncludingConnectionCloseHeader() throws IOException {
+    try (Socket slowRequestConnection = new Socket("localhost", dynamicPort.getNumber())) {
+      sendRequest(slowRequestConnection, SLOW_PROCESSING_ENDPOINT);
+
+      // Stop mule in parallel.
+      stopMuleExecutor.execute(new MuleContextStopper());
+      assertContextIsStopping(SMALL_TIMEOUT_MILLIS);
+
+      // Response is ok, but connection close header is added.
+      String slowRequestResponse = getResponse(slowRequestConnection);
+      assertResponse(slowRequestResponse, true);
+      assertThat(slowRequestResponse, containsString("Connection: close"));
+
+      // After that, the connection is no longer usable.
+      sendRequest(slowRequestConnection, FAST_PROCESSING_ENDPOINT);
+      slowRequestResponse = getResponse(slowRequestConnection);
+      assertResponse(slowRequestResponse, false);
+    }
+  }
+
+  @Test
+  public void serverIsStoppedWhenPersistentConnectionsAreClosed() throws IOException {
+    try (Socket idlePersistentConnection = generateIdlePersistentConnection()) {
+      // Stop mule in parallel.
+      stopMuleExecutor.execute(new MuleContextStopper());
+      assertContextIsStopping(SMALL_TIMEOUT_MILLIS);
+
+      // There is a persistent connection open, so muleContext doesn't stop during the shutdown timeout.
+      assertThat(muleContext.isStopped(), is(false));
+
+      idlePersistentConnection.close();
+
+      // When the persistent connection is closed, muleContext stops.
+      assertContextHasStopped(SMALL_TIMEOUT_MILLIS);
+    }
+  }
+
+  @Test
+  public void onlyOneRequestIsValidUsingAnOldPersistentDuringShutdown() throws IOException, InterruptedException {
+    Socket persistentConnection1 = generateIdlePersistentConnection();
+    Socket persistentConnection2 = generateIdlePersistentConnection();
+    Socket persistentConnection3 = generateIdlePersistentConnection();
+
+    // Stop mule in parallel.
+    stopMuleExecutor.execute(new MuleContextStopper());
```

Comment:
if you use submit instead, in the case the task fails you can get the failure from the returned `Future`